### PR TITLE
adding fileutil

### PR DESCRIPTION
**Pull request by tvoran**
Adds a fileutil package, mainly for the purpose of sharing the caching file reader from https://github.com/hashicorp/vault-plugin-auth-kubernetes/blob/main/caching_file_reader.go

### CODE REVIEW

File: fileutil/caching_file_reader.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	r.l.Lock()
+	defer r.l.Unlock()
+
+	buf, err := ioutil.ReadFile(r.path)
+	if err != nil {
+		return "", err
+	}
+	r.cache = cachedFile{
+		buf:    string(buf),
+		expiry: now.Add(r.ttl),
```
Comment (tomhjp) [inline fileutil/caching_file_reader.go:64]: It's very unlikely, and would require exceedingly high load, but we could theoretically spend an arbitrary amount of time waiting for the write lock at line 55, and so we could end up storing an expiry time that's already in the past, or at least more stale than it should be.

```suggestion
		expiry: r.currentTime().Add(r.ttl),
```

**Review by tomhjp — APPROVED** [review]

File: fileutil/caching_file_reader.go (hunk 1 above)
Comment (tvoran) [reply fileutil/caching_file_reader.go]: Agreed, thanks!

**Review by tomhjp — DISMISSED** [review]

File: fileutil/caching_file_reader.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	r.l.RUnlock()
+	if now.Before(cache.expiry) {
+		return cache.buf, nil
+	}
+
+	// Slow path: read the file from disk.
+	r.l.Lock()
+	defer r.l.Unlock()
+
+	buf, err := ioutil.ReadFile(r.path)
```
Comment (swenson) [inline fileutil/caching_file_reader.go:58]: nit: Since we're on Go 1.16, can we use `os.ReadFile`?

**Review by swenson — APPROVED** [review]
LGTM; small nits about removing `ioutil`

File: fileutil/caching_file_reader.go (hunk 2 above)
Comment (tvoran) [reply fileutil/caching_file_reader.go]: Yep, fixed in [5b61c29](https://github.com/hashicorp/go-secure-stdlib/pull/38/commits/5b61c292446dcaa6f7b50783fc501a72e50bd94a)

File: fileutil/caching_file_reader_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
+package fileutil
+
+import (
+	"io/ioutil"
```
[the remark this answers is not in the document]
Comment (tvoran) [reply fileutil/caching_file_reader_test.go]: Fixed in [5b61c29](https://github.com/hashicorp/go-secure-stdlib/pull/38/commits/5b61c292446dcaa6f7b50783fc501a72e50bd94a)

**Review by calvn — APPROVED** [review]

**Review by tomhjp — APPROVED** [review]

**Pull request merged**